### ISSUE TITLE
We can hook sleep.target, but there is no corresponding wakeup.target

### ISSUE THREAD

**Issue by sourcejedi** [systemd/systemd#6364]
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with

c1edab7

### Used distribution

Fedora Linux 25

### RFE

Services can be started at suspend (to ram, disk, or both) by `WantedBy=sleep.target` `Before=sleep.target`.  However there is no equivalent for resume.  Note that `After=sleep.target` does *not* start services on resume.  <del>Ordering dependencies on `sleep.target` have no significance</del>.

This is incoherent.  There are several use cases for hooking into resume (seen in unix.stackexchange.net questions), and the documentation does not suggest any reason to discourage it.  Instead, you have to hook in to each type of suspend individually.

I think this could be addressed by adding `wakeup.target` to match `sleep.target`, with systemd-`suspend.service` etc modified to include `Requires=wakeup.target`, `Before=wakeup.target`.

**Comment by arvidjaar** [comment]
> Note that After=sleep.service does not start services on resume.

Even if you add `WantedBy=sleep.service`?

**Comment by sourcejedi** [comment]
No, of course it didn't :).  Because `systemd-suspend.service` which launches `/usr/lib/systemd/systemd-sleep suspend` is ordered `After=sleep.service`.

**Comment by arvidjaar** [comment]
@sourcejedi What is `sleep.service`? Are you sure you do not mean `sleep.target`?

You can make your service to be started on resume by ordering it explicitly against `systemd-suspend.service` & Co. If it proves useful I suppose it could be possible to add standard target for it.

**Comment by sourcejedi** [comment]
To my mind, the problem is the lack of `wakeup.target` heavily implies that you can use `WantedBy=sleep.target` + `After=sleep.target`. Just like you assumed.

The main use is this:

* https://unix.stackexchange.com/questions/347695/to-configure-hard-drive-apm-levels-and-maintain-them-after-suspend-resume
* https://unix.stackexchange.com/questions/80437/how-can-i-run-an-hdparm-command-after-boot-and-resume-on-fedora-19
* https://forums.opensuse.org/showthread.php/510577-How-to-execute-after-local-after-resume-from-suspend-hibarnate

etc.  This works really nice because the same service unit can handle both boot and resume.  And when you have a simple command to run, service units are actually simpler than writing a shell script hook.  You don't have to know which numbered argument is for suspend/resume v.s. disk/mem/both, and write code to check it.

On stackexchange there was one other case, working around a backlight bug.

* https://unix.stackexchange.com/questions/124212/writing-a-systemd-service-to-be-executed-at-resume/337178#337178

I've seen people on stackexchange get worried about how `/usr/lib/systemd/system-sleep` hook scripts are 

1. described "should be considered hacks".  (Earlier there was a unhelpfully vague mention that there are better ways to do it, though this is now replaced with a reference to inhibitors.  Hopefully the new text is clearer for people to understand.)
2. placed in `/usr`, apparently breaking the FHS.

After consideration, it's not that I think I need to avoid those on my own system.  But I think we could save people from having to think about this so hard, and it's a pretty straightforward solution.

**Comment by sourcejedi** [comment]
You're right, I had written `sleep.service` sometimes when I meant `sleep.target`.

**Comment by arvidjaar** [comment]
@sourcejedi 
> Just like you assumed.

Ehh ... I actually initially assumed that your `sleep.service` was generic reference to `systemd-suspend.service` et al.

**Comment by sourcejedi** [comment]
Ah, wrong assumption on my part.

**Comment by poettering** [comment]
Why wouldn't it suffice to do:

```
[Unit]
WantedBy=sleep.target
After=systemd-suspend.service systemd-hybrid-sleep.service systemd-hibernate.service
```

?

**Comment by sourcejedi** [comment]
That should also work, as I understand it.  You could say my first issue is the documentation.  Adding `wakeup.target` is the most straightforward way I can think of, to make the documentation useful.

We advertise `WantedBy=sleep.target` as a way to "hook units into the sleep state logic".  But the documentation does not provide enough information to use it to run commands after resume.  _Or before suspend_.  Because 1) it doesn't give any example and 2) it doesn't mention that any ordering constraints are applied to `sleep.target`.

I'm unclear what utility there would be at all, being able to configure a unit which is triggered by a sleep transaction, but is not ordered w.r.t kernel suspend/resume.  But that's all the docs tell you about `sleep.target`.

An individual unit which references both `sleep.target` and the 3 individual services seems unnecessarily confusing to readers.  Why not write it as `systemd-suspend.service systemd-hybrid-sleep.service systemd-hibernate.service` for both `WantedBy=` and `After=`?  However people who see a resume unit which uses the 3 individual services, and doesn't use the `sleep.target` mentioned in the documentation, will also be puzzled to some extent.

People who may not have written a unit before are understandably attempting to use `WantedBy=sleep.target` `After=sleep.target`.  They become frustrated and mystified as to why it doesn't work as expected, but it does when they experiment with your suggestion.

People know that hooking resume is a Thing.  Not just because they have a desire themselves, but because it's supported by `systemd-sleep` (and historically `pm-utils`).  So the way `systemd` has a `sleep.target` but not a `wakeup.target` is actively confusing people, into thinking that `After=sleep.target` could do something useful.

(And they like to use units in preference to `/usr/lib/systemd/systemd-sleep/`.  Because both that location and the docs are us deliberately nudging people away from using it.  And for the hdparm APM mess, one simple service unit can handle running the command on both boot and resume).

**Comment by mbiebl** [comment]
> Why wouldn't it suffice to do:
> 
> [Unit]
> WantedBy=sleep.target
> After=systemd-suspend.service systemd-hybrid-sleep.service systemd-hibernate.service

This seems rather convoluted and counter-intuitive. I'd prefer if we had an explicit target for this use case.

For fun, I just checked the Debian archive:
https://codesearch.debian.net/search?q=WantedBy%3Dsleep.target

Seems people get it wrong with the currently available logic/documentation, see
http://sources.debian.net/src/tlp/0.9-3/tlp-sleep.service.in/?hl=19#L19 which tries to run something on resume via ExecStop or
http://sources.debian.net/src/syncthing/0.14.31%2Bdfsg1-1/etc/linux-systemd/system/syncthing-resume.service/?hl=11#L11 which tries to run something on resume via After=sleep.target.

Our documentation or existing mechanisms apparently failed to convey how to do it "properly".

**Comment by sourcejedi** [comment]
I was imagining the changes to the docs would not necessarily spell all this out.  Hopefully just mentioning `wakeup.target` next to `sleep.target` would imply most of it.  (I expect we'd be having the same discussion if I'd submitted that as a PR).

**Comment by sourcejedi** [comment]
@mbiebl I'm probably projecting, but I also can't help wondering if this would have got the Debian developers to not use the discouraged `/usr/lib/systemd/systemd-sleep/` in the hdparm package :).

**Comment by mbiebl** [comment]
@sourcejedi most likely yes.
See https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=744753#158
and https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=744753#241

**Comment by mbiebl** [comment]
@sourcejedi fwiw, this is an old discussion, see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=744753#25

**Comment by mbiebl** [comment]
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=744753#124 has another example of users getting it wrong trying to run something on resume.

**Comment by sourcejedi** [comment]
> Seems people get it wrong with the currently available logic/documentation, see
http://sources.debian.net/src/tlp/0.9-3/tlp-sleep.service.in/?hl=19#L19 which tries to run something on resume via ExecStop

crap.  I was asked on #6444 if I could document an example of both sleep and resume units, and this was the only practical case I could think of for the sleep side.  But looking at it, I think this ExecStop approach should be working well.

It actually seems a more explicitly defensive approach, in case `systemd-hibernate.service` could fail.  After failure, `tlp` would still want to run e.g. the `resume_drivebay` function to undo the effect of `suspend_drivebay`.  The old `pm-utils` worked like this (`systemd-sleep` hooks do not, which seems annoying).

I tested an equivalent of wakeup.target when systemd-sleep.service fails.  wakeup.target seems to run still, as would be desired by `tlp`.  But I don't understand it enough to be sure... e.g. will that still happen when someone works out how to fix issue 6419 ?  If wakeup.target might not do the right thing there, it would not be great to add it and tell people to start using it.

**Comment by jackm** [comment]
I've read through this issue and many other sites posing the same question however I am still confused as to what is the recommended way to run a service _after_ the system has resumed.

I do see that the `wakeup.target` commit was made however I am using systemd version 229 which does not have the `wakeup.target` so I'd have to do it the "old way".

Could someone give me a concise answer on how to ensure that a service runs _after_ the system has:
1. resumed from sleep of any kind, including _suspend_, _hibernate_, _hybrid-sleep_, etc
1. resumed from suspend only
1. resumed from hibernate only

**Comment by theGeekyLad** [comment]
I need to show a GUI before my system suspends and the simple `Before=sleep.target` under unit and `WantedBy=sleep.target` under install runs my GUI trigger script after the screen goes off. :(

**Comment by moritzschaefer** [comment]
Since this is still an "open" issue and many people are searching for an answer (see the 12 likes on @jackm's comment), the solution, as I understand, is provided by @poettering in his [answer](https://github.com/systemd/systemd/issues/6364#issuecomment-316647050) above:

```
[Unit]
WantedBy=sleep.target
After=systemd-suspend.service systemd-hybrid-sleep.service systemd-hibernate.service
```

**Comment by drws** [comment]
We're over the 6-year mark for a simple, obvious and even already-solved issue whose solution should be merged even before the first distro picked systemd. And counting, apparently years, given latest activity.

**Comment by qwelias** [comment]
@poettering @moritzschaefer does it still work for you if `Conflicts=sleep.target`?
It does not for me, need to stop the service before sleep and resume on wake

**Comment by moritzschaefer** [comment]
Yes, it is still working for me so far, but haven't updated in a while. Not sure what 23.11 will bring ;)

**Comment by qwelias** [comment]
Seems like it's not so straightforward when adding that snippet to an existing service then

**Comment by drws** [comment]
The point of this issue should not be to work around the lack of `wakeup.target` or to manually implement it, but to solve the issue **for real**.

**Comment by HaleTom** [comment]
Here's how you can actually have a `resume.target`:

`/etc/systemd/system/resume.target`:

```
# Started by ExecStop= or OnSuccess= in resume.service
# See sleep.target in systemd.special(7)
# Intended to be hooked in by other units' "RequiredBy=resume.target"

[Unit]
Description=Resume from sleep
DefaultDependencies=no
StopWhenUnneeded=yes
```

```
# Start resume.target after waking up
[Unit]
Description=Schedule resume.target startup after resume
DefaultDependencies=no
RefuseManualStart=yes
StopWhenUnneeded=yes
# Before= shutdown (ExecStop) tasks are done AFTER those of the specified dependency
Before=sleep.target
OnSuccess=resume.target

[Service]
Type=oneshot
RemainAfterExit=yes
# There needs to be at least one ExecStart= or ExecStop=
ExecStart=/bin/true
# Using OnSuccess= above instead
# ExecStop=/usr/bin/systemctl start resume.target

[Install]
# Make it a passive member of every type of sleep
WantedBy=sleep.target
```

I'm still undecided whether to prefer the `OnSuccess=` without running another process (`systemctl`), or accepting that `OnSuccess=` includes when `sleep.target` goes inactive in very unlikely edge cases.

Any comments on this choice?

**Comment by sourcejedi** [comment]
In an installable package or re-usable ansible role, I'd prefer the [documented method](https://github.com/sabily-os/systemd/commit/edecb56216c910b36980fcf590b385ede413392b).  (It's what I'll use, if I ever get back to my [original use case](https://github.com/sourcejedi/ansible-hdparm-noapm)).

For a sysadmin who prefers to add `resume.target` (why?), I suggest either 1) accept running `systemctl`, because that both follows the exact pattern documented, and helps mark it as an "unusual" setup.  Or 2) can't you just use `OnSuccess=` in a drop-in file, to avoid the extra unit and lines of code?

```
# /etc/systemd/system/sleep.target.d/override.conf
[Unit]
OnSuccess=resume.target
```
